packages/frontend-app-api/src/wiring/createSpecializedApp.tsx

```diff
@@ -95,21 +97,47 @@ function deduplicateFeatures(
 
 // Helps delay callers from reaching out to the API before the app tree has been materialized
 class AppTreeApiProxy implements AppTreeApi {
-  #safeToUse: boolean = false;
+  #routeInfo?: RouteInfo;
 
-  constructor(private readonly tree: AppTree) {}
+  constructor(
+    private readonly tree: AppTree,
+    private readonly appBasePath: string,
+  ) {}
 
-  getTree() {
-    if (!this.#safeToUse) {
+  private checkIfInitialized() {
+    if (!this.#routeInfo) {
       throw new Error(
         `You can't access the AppTreeApi during initialization of the app tree. Please move occurrences of this out of the initialization of the factory`,
       );
     }
+  }
+
+  getTree() {
+    this.checkIfInitialized();
+
     return { tree: this.tree };
   }
 
-  initialize() {
-    this.#safeToUse = true;
+  getNodesByRoutePath(sourcePath: string): { nodes: AppNode[] } {
+    this.checkIfInitialized();
+
+    let path = sourcePath;
+    if (sourcePath.startsWith(this.appBasePath)) {
+      path = sourcePath.slice(this.appBasePath.length);
+    }
+
+    const matchedRoutes = matchRoutes(this.#routeInfo!.routeObjects, path);
+
+    const matchedAppNodes =
+      matchedRoutes
+        ?.filter(routeObj => !!routeObj.route.appNode)
+        .map(routeObj => routeObj.route.appNode!) || [];
```

Comment:
flipping this around simplifies it a bit, i.e. `.map(...).filter(Boolean)`

Reply:
Had to revert that since making TS happy would require making it less simple